integration/oasis/chargingstrategy/chargingstrategy.go

```diff
@@ -11,24 +15,85 @@ func NewFakeChargingStrategyCreator(maxEnergyLevel float64) *fakeChargingStrateg
 	}
 }
 
+// @todo:
+// - Influence of returning candidate with no charge time and additional energy
 // CreateChargingStrategies returns different charging strategy
-// Initial implementation: 1 hour charge for 60% of max energy,
-//                         2 hour charge for 80%
-//                         4 hour charge for 100%
-func (f *fakeChargingStrategyCreator) CreateChargingStrategies() []ChargingStrategy {
-
-	return []ChargingStrategy{
-		ChargingStrategy{
-			ChargingTime:   3600,
+func (f *fakeChargingStrategyCreator) CreateChargingStrategies() []ChargingStatus {
+	return []ChargingStatus{
+		ChargingStatus{
 			ChargingEnergy: f.maxEnergyLevel * 0.6,
 		},
-		ChargingStrategy{
-			ChargingTime:   7200,
+		ChargingStatus{
 			ChargingEnergy: f.maxEnergyLevel * 0.8,
 		},
-		ChargingStrategy{
-			ChargingTime:   14400,
+		ChargingStatus{
 			ChargingEnergy: f.maxEnergyLevel,
 		},
 	}
 }
+
+// Fake charge strategy
+// From empty energy:
+//                    1 hour charge to 60% of max energy
+//                    2 hour charge to 80%, means from 60% ~ 80% need 1 hour
+//                    4 hour charge to 100%, means from 80% ~ 100% need 2 hours
+func (f *fakeChargingStrategyCreator) EvaluateCost(arrivalEnergy float64, targetState ChargingStatus) ChargingCost {
+	sixtyPercentOfMaxEnergy := f.maxEnergyLevel * 0.6
+	eightyPercentOfMaxEnergy := f.maxEnergyLevel * 0.8
+	noNeedCharge := ChargingCost{
+		Duration: 0.0,
+	}
+
+	if arrivalEnergy > targetState.ChargingEnergy ||
+		floatEquals(targetState.ChargingEnergy, 0.0) {
+		return noNeedCharge
+	}
+
+	totalTime := 0.0
+	currentEnergy := arrivalEnergy
+	if arrivalEnergy < sixtyPercentOfMaxEnergy {
+		energyNeeded4Stage1 := sixtyPercentOfMaxEnergy - arrivalEnergy
+		totalTime += energyNeeded4Stage1 / sixtyPercentOfMaxEnergy * 3600.0
+		currentEnergy = sixtyPercentOfMaxEnergy
+	}
+
+	if floatEquals(targetState.ChargingEnergy, sixtyPercentOfMaxEnergy) {
+		return ChargingCost{
+			Duration: totalTime,
+		}
+	}
+
+	if arrivalEnergy < eightyPercentOfMaxEnergy {
+		energyNeeded4Stage2 := eightyPercentOfMaxEnergy - currentEnergy
+		totalTime += energyNeeded4Stage2 / (eightyPercentOfMaxEnergy - sixtyPercentOfMaxEnergy) * 3600.0
+		currentEnergy = eightyPercentOfMaxEnergy
+	}
+	if floatEquals(targetState.ChargingEnergy, eightyPercentOfMaxEnergy) {
+		return ChargingCost{
+			Duration: totalTime,
+		}
+	}
+
+	if arrivalEnergy < f.maxEnergyLevel {
+		energyNeeded4Stage3 := f.maxEnergyLevel - currentEnergy
+		totalTime += energyNeeded4Stage3 / (f.maxEnergyLevel - eightyPercentOfMaxEnergy) * 7200.0
+	}
+
+	if floatEquals(targetState.ChargingEnergy, f.maxEnergyLevel) {
+		return ChargingCost{
+			Duration: totalTime,
+		}
+	}
+
+	glog.Fatalf("Invalid charging state %#v\n", targetState)
+	return noNeedCharge
+}
+
+var epsilon float64 = 0.00000001
+
+func floatEquals(a, b float64) bool {
```

Comment:
We really need to find/implement a `floatEquals` package. There're too many places want to use it.

Reply:
Yes, and `location`

Reply:
How about add `common` folder in `integration`
- For floatEquals, we add into `util.go`.  Search this from other repository, one sample is [`go-gl/mathgl/mathgl32/util`](https://github.com/go-gl/mathgl/blob/5ab0e04e1f55a10f3a71b1fccc60ca2ee9a0bc02/mgl32/util.go#L40:6)
- For `Location` could put which into `type` folder

Reply:
Not only `oasis` need it, but also other packages. What if create a `integration/util` package for the `floatEquals`, and  a `integration/map` package for the `Location`? The `Location` here is for general purpose, if some special usage(e.g., `Coordinate` in `osrm`) just create their owns.

Reply:
Ok, will put it to `integration`